x/tokenomics/keeper/token_logic_modules.go

```diff
@@ -0,0 +1,443 @@
+		return tokenomicstypes.ErrTokenomicsApplicationNotFound
+	}
+
+	// Retrieve the on-chain staked supplier record
+	supplier, foundSupplier := k.supplierKeeper.GetSupplier(ctx, supplierAddr.String())
+	if !foundSupplier {
+		logger.Warn(fmt.Sprintf("supplier for claim with address %q not found", supplierAddr))
+		return tokenomicstypes.ErrTokenomicsSupplierNotFound
+	}
+
+	service, ok := k.serviceKeeper.GetService(ctx, sessionHeader.Service.Id)
+	if !ok {
+		return tokenomicstypes.ErrTokenomicsServiceNotFound.Wrapf("service with ID %q not found", sessionHeader.Service.Id)
+	}
+
+	// Determine the total number of tokens that'll be used for settling the session.
+	// When the network achieves equilibrium, this will be the mint & burn.
+	settlementCoin, err = k.computeUnitsToCoin(ctx, claimComputeUnits, &service)
+	if err != nil {
+		return err
+	}
+
+	// Retrieving the relay mining difficulty for the service at hand
+	relayMiningDifficulty, found := k.GetRelayMiningDifficulty(ctx, service.Id)
+	if !found {
+		numRelays, err := claim.GetNumRelays()
+		if err != nil {
+			return err
+		}
+		logger.Warn(fmt.Sprintf("relay mining difficulty for service %q not found. Using default difficulty", service.Id))
+		relayMiningDifficulty = tokenomicstypes.RelayMiningDifficulty{
+			ServiceId:    service.Id,
+			BlockHeight:  sdk.UnwrapSDKContext(ctx).BlockHeight(),
+			NumRelaysEma: numRelays,
+			TargetHash:   prooftypes.DefaultRelayDifficultyTargetHash,
+		}
+	}
+
+	// Helpers for logging the same metadata throughout this function calls
+	logger = logger.With(
+		"compute_units", claimComputeUnits,
+		"session_id", sessionHeader.GetSessionId(),
+		"service_id", sessionHeader.GetSessionId(),
+		"supplier", supplier.Address,
+		"application", application.Address,
+	)
+	logger.Info(fmt.Sprintf("About to start processing TLMs for (%d) compute units equaling to (%s) coins", claimComputeUnits, settlementCoin))
+
+	// Execute all the token logic modules processors
+	for tlm, tlmProcessor := range tokenLogicModuleProcessorMap {
+		logger.Info(fmt.Sprintf("Starting to execute TLM %q", tlm))
+		if err := tlmProcessor(k, ctx, &service, &application, &supplier, settlementCoin, &relayMiningDifficulty); err != nil {
+			return err
+		}
+		logger.Info(fmt.Sprintf("Finished executing TLM %q", tlm))
+	}
+
+	// Update the application's on-chain record
+	k.applicationKeeper.SetApplication(ctx, application)
+	logger.Info(fmt.Sprintf("updated on-chain application record with address %q", application.Address))
+
+	// Update the suppliers's on-chain record
+	k.supplierKeeper.SetSupplier(ctx, supplier)
+	logger.Info(fmt.Sprintf("updated on-chain supplier record with address %q", supplier.Address))
+
+	// Update isSuccessful to true for telemetry
+	isSuccessful = true
+	return nil
+}
+
+// TokenLogicModuleRelayBurnEqualsMint processes the business logic for the RelayBurnEqualsMint TLM.
+func (k Keeper) TokenLogicModuleRelayBurnEqualsMint(
```

Comment:
Could we have each TLM processor live in its own file?

Reply:
That's not a bad idea. 

This might undergo another round of changes (smaller ones) so I'll hold off until we see how that looks.